[Communication - Telephony Service] Trunk status API

dzeliar opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Changelog
Add a changelog entry for this PR by answering the following questions:
  1. What's the purpose of the update?
      - [ ] new service onboarding
      - [x] new API version
      - [ ] update existing version for new feature
      - [ ] update existing version to fix swagger quality issue in s360
      - [ ] Other, please clarify
  2. When are you targeting to deploy the new service/feature to public regions? Please provide the date or, if the date is not yet available, the month.
  3. When do you expect to publish the swagger? Please provide date or, the the date is not yet available, the month.
  4. If updating an existing version, please select the specific language SDKs and CLIs that must be refreshed after the swagger is published.
      - [ ] SDK of .NET (need service team to ensure code readiness)
      - [ ] SDK of Python
      - [ ] SDK of Java
      - [x] SDK of Js
      - [ ] SDK of Go
      - [ ] PowerShell
      - [ ] CLI
      - [ ] Terraform
      - [ ] No refresh required for updates in this PR

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of "no breaking changes"
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist 

> **Applicability**: :warning: 
>
> If your changes encompass only the following scenarios, you should SKIP this section, as these scenarios do not require ARM review.
> - Change to data plane APIs 
> - Adding new properties 
> - All removals

Otherwise your PR may be subject to ARM review requirements. Complete the following:
- [ ] Check this box if any of the following appy to the PR so that the label "ARMReview" and "WaitForARMFeedback" will be added by bot to kick off ARM API Review.  Missing to check this box in the following scenario may result in delays to the ARM manifest review and deployment.
  - Adding a new service
  - Adding new API(s)
  - Adding a new API version
    -[ ] To review changes efficiently, ensure you are using OpenAPIHub to initialize the PR for adding a new version. More details, refer to the [wiki](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/208/OpenAPI-Hub-Adding-new-API-version).

- [ ] Ensure you've reviewed following [guidelines](https://aka.ms/rpguidelines) including [ARM resource provider contract](https://github.com/Azure/azure-resource-manager-rpc) and [REST guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md). Estimated time (4 hours). This is required before you can request review from ARM API Review board.

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If any of the following scenarios apply to the PR, request approval from the Breaking Change Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 
- [ ] Removing API(s) in a stable version
- [ ] Removing properties in a stable version
- [ ] Removing API version(s) in a stable version
- [ ] Updating API in a stable or public preview version with Breaking Change Validation errors
- [ ] Updating API(s) in public preview over 1 year (refer to [Retirement of Previews](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37683/Retirement-of-Previews))

**Action**: to initiate an evaluation of the breaking change, create a new intake using the [template for breaking changes](https://aka.ms/Breakingchangetemplate). Addition details on the process and office hours are on the [Breaking change Wiki](https://dev.azure.com/msazure/AzureWiki/_wiki/wikis/AzureWiki.wiki/37684/Breaking-Changes).

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

fix https://github.com/Azure/azure-rest-api-specs/issues/19376
fix https://github.com/Azure/azure-rest-api-specs/issues/19376

dzeliar committed:
trunkstatus 2022-07-01 preview

dzeliar committed:
fix nextLink

weidongxu-microsoft commented:
@dzeliar 
Please see https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship

dzeliar commented:
As I'll be OOF for the next two weeks, @nikuklic will address any comments

jiriburant commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          }
+        ],
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
```
The paging was not fixed properly, cannot generate code from this specification. Consider adding:
"itemName": "values"

jiriburant commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          }
+        ],
+        "x-ms-pageable": {
+          "nextLinkName": "nextLink"
```
The paging was not fixed properly, cannot generate code from this specification. Consider adding:
"itemName": "values"

dzeliar replied:
Removed pageable completely as we expect to have on average just a few records, max is 250

jiriburant commented on specification/communication/data-plane/TrunkStatus/readme.md:
```diff
@@ -0,0 +1,54 @@
+
+## CSharp
+
+See configuration in [readme.csharp.md](./readme.csharp.md)
```
Add link to typescript, not csharp.

dzeliar replied:
Updated

jiriburant commented on specification/communication/data-plane/TrunkStatus/readme.md:
```diff
@@ -0,0 +1,54 @@
+
+## CSharp
+
+See configuration in [readme.csharp.md](./readme.csharp.md)
```
Add link to typescript, not csharp.

dzeliar committed:
remove pageable, fix sdk typo

weidongxu-microsoft commented:
There is no update to https://github.com/Azure/azure-rest-api-specs/issues/19349

Let reviewer know after meeting is scheduled and discussion is added (usually Mark's reply in issue).

dzeliar commented:
Hello @weidongxu-microsoft , our review has been scheduled for today (Wed 29/06/2022)

dzeliar committed:
add port to trunk properties

dzeliar committed:
replace by health endpoint

dzeliar committed:
fix overallHealth property

dzeliar committed:
fix enums

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+    },
+    "version": "2022-07-01-preview1"
+  },
+  "paths": {
```
Overall, this new API seems a bit too narrow to me, especially in the following context:

`/sip` already has a GET method. The response to that includes a map of `trunks`. The map there contains trunk FQDN (as the key of the map) and the `sipSignalingPort`.
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/SipRouting/preview/2021-05-01-preview/communicationservicessiprouting.json

Now a part of the information about the trunk is split to this separate API under `sip/trunkHealth`.

Two desires:
* Keep information about trunks in one place instead of two.
* Avoid a narrow api path that only stores health information for trunks.

There are a few options to meet these desires:
* Do not introduce the `trunkHealths` collection, instead, expand the `trunks` map under `/sip` and depend on JSON merge-path etc.
  * Should only do this if the set of `trunkHealths` changes a lot, and/or is limited in size.
* Introduce `/sip/trunks` and move the FQDN and `sipSignalingPort` in there. Drop the map under `/sip`
* Introduce `/sip/trunks` to contain static information (like `sipSignalingPort` and `fqdn`) and `sip/trunks/<fqdn>/health` to get the dynamic health for a trunk. The problem here is that it's harder to get the health for *all* trunks.

There is the extra complication that health information is changed by the server, while `sipSignalingPort` is changed by the client (?). This could be confusing. The last option (and the current proposal) mitigates this by keeping the two separate.

--

In any case, splitting the trunk information across a map and a collection seems weird because maps and collections fit different requirements:
e.g, collections must support pagination, with the assumption that collection can grow to be large. In this case, having related information in a map becomes a problem.

mbakalov replied:
FYI related suggestion from a previous review of this api: #19376

mbakalov replied:
I recall there was also a suggestion to use `expand`: `/sip/trunks?expand=healthStatus` or `/sip?expand=healthStatus` - did we consider this?

dzeliar replied:
As agreed on internal review, we will expand `trunks` map under `/sip?expand=trunkHealth`

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          "200": {
+            "description": "Ok.",
+            "schema": {
+              "$ref": "#/definitions/ListOfTrunksHealth"
```
Avoid including type in the model name (`ListOf`).

You can follow chat's example and call this `TrunkHealthCollection`, or simply `TrunkHealths` (pluralized)

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/Chat/stable/2021-09-07/communicationserviceschat.json#L1238

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+      "properties": {
+        "value": {
+          "description": "Represents an array of SIP trunks health.",
+          "maxLength": 250,
```
This should be `maxItems`?
https://datatracker.ietf.org/doc/html/draft-bhutton-json-schema-validation-01#section-6.4.1

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+      "description": "Health of SIP trunks.",
+      "type": "object",
+      "properties": {
+        "value": {
```
The guidelines are fairly clear that we should paginate all collections.

I do see below that we're stating the size of this collection is limited to 250. How hard is this limit? If it's set by laws of physics / a fairly stable standard, we don't need to paginate. If it's based more on current requirements, we should paginate.

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          "enum": [
+            "NoData",
+            "OK",
+            "CertExpiring",
```
Food for thought: Should this somehow include _when_ the certificate is expiring?

dzeliar replied:
We can expand heath with multiple properties in future, this is not requested in the current spec

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          }
+        },
+        "ping": {
+          "description": "The status of options message send by SBC",
```
grammar nit: s/send/sent/

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          "enum": [
+            "NoData",
+            "Active",
+            "InactiveNoRecentPings",
```
I'm somewhat uncomfortable with enums that have combinatorial variants: `*PingsAndCalls`. How do we expect this enum to grow in the future? Do we risk blowing up the number of variants? Are clients likely to always have check for multiple related variants in their code (e.g. `if (overallStatus == 'InactiveNoRecentPings' || overallStatus == 'InactiveNoRecentCalls' || overallStatus == 'InactiveNoRecentPingsAndCalls'`)

This smells like it could be split into two separate fields: The status is limited to `NoData | Active | Inactive`. And then for inactive trunks, there are optional reasons `inactiveReasons = [](NoRecentCalls | NoRecentPings)`

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+            "NoData",
+            "OK",
+            "PingError",
+            "PingExpired"
```
Food for thought: What is the difference between `NoData` and `PingExpired` from Contoso POV? Can they trigger a ping in the latter case? Otherwise, is there anything different they would do for these two cases? Can we get away with just `NoData` (no reason to tell them _why_ there is no data)

dzeliar replied:
PingError is returned when ping status is not one of the above. Not NoData, OK or Expired. NoData (we don't have any records) is different compared to PingExpired (we have old data but unable to get up-to-date status)

prprabhu-ms commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          "description": "The status of options message send by SBC",
+          "type": "string",
+          "enum": [
+            "NoData",
```
`NoData` is an unconventional default value for an enum.

A quick search suggests that `unknown` is a more conventional choice (also in my experience):

https://github.com/Azure/azure-rest-api-specs/search?q=Unknown
https://github.com/Azure/azure-rest-api-specs/search?q=Undetermined
https://github.com/Azure/azure-rest-api-specs/search?q=NoData

DominikMe commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+          "description": "The overall status of SBC",
+          "type": "string",
+          "enum": [
+            "NoData",
```
Ideally we stick to lower camelCase and not PascalCase for everything including enum values.

dzeliar committed:
update enum and add descriptive properties

mbakalov commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,212 @@
+    "version": "2022-07-01-preview1"
+  },
+  "paths": {
+    "/sip/trunksHealth": {
```
Or maybe `/trunkHealthStatuses` and `/trunkHealthStatuses/{fqdn}`?

dzeliar committed:
enums to camelcase

DominikMe commented on specification/communication/data-plane/TrunkStatus/preview/2022-07-01-preview1/communicationservicestrunkstatus.json:
```diff
@@ -0,0 +1,327 @@
+{
+  "swagger": "2.0",
+  "info": {
+    "title": "Trunk health Service",
```
This service seems fairly minimal in scope, why didn't we integrate it with the SipRouting service? The overhead of another service and SDK isn't small and I would assume as a developer I can't use TrunkHealth if I don't also use SipRounting?

DominikMe replied:
Could the health data inlined with the rest of the Trunk properties in the SipRouting service?

dzeliar committed:
remove trunkstatus folder

dzeliar committed:
add health to trunk properties

dzeliar committed:
create new preview

dzeliar committed:
address review comments

dzeliar committed:
fix example

dzeliar committed:
the latest api version swaggers are in the default tag

dzeliar committed:
remove unused response codes from example

dzeliar committed:
fix prettier

dzeliar committed:
update python and ruby tag to the latest version

kdruzycki commented on specification/communication/data-plane/SipRouting/preview/2022-09-01-preview/communicationservicessiprouting.json:
```diff
@@ -0,0 +1,454 @@
+  },
+  "parameters": {
+    "ExpandSipConfiguration": {
+      "name": "expand",
```
Maybe you could consider naming this `$expand` instead of `expand`?
That would be in tact with OData standard https://docs.microsoft.com/en-us/aspnet/web-api/overview/odata-support-in-aspnet-web-api/using-select-expand-and-value
Also it would be congruent with other usages of `$expand` in this repository https://github.com/Azure/azure-rest-api-specs/search?q=%22%5C%22name%5C%22%3A+%5C%22%24expand%5C%22%22&type=code

dzeliar replied:
Yes those should have never been there

kdruzycki commented on specification/communication/data-plane/SipRouting/preview/2022-09-01-preview/communicationservicessiprouting.json:
```diff
@@ -0,0 +1,454 @@
+            "schema": {
+              "$ref": "#/definitions/SipConfigurationExpanded"
+            }
+          },
```
Error code responses 415, 422, 500 have been removed compared to the previous version...
https://github.com/Azure/azure-rest-api-specs/blob/main/specification/communication/data-plane/SipRouting/preview/2021-05-01-preview/communicationservicessiprouting.json#L80
Is there a reason for that?

dzeliar committed:
Unify naming SBC->Trunk, move Extended to beginning of Trunk and SipConfiguration to suggest inheritance

dzeliar commented:
Hello everyone,


on yesterdays board review of SipRouting - Trunk health, [Issue #19376](https://github.com/Azure/azure-rest-api-specs/issues/19376), we agreed on the following steps:

1, Rename SipConfigurationExpanded & TrunkExpanded -> ExpandedSipConfiguration & ExpandedTrunk
2, Unify naming Trunk vs. SBC vs. SessionBorderController. After review with our PM we unified naming and now we use only Trunk.

All changes we pushed and the PR is now ready to merge. In order to do so I need an approval from API Stewardship Board.
Please, @mikekistler , @JeffreyRichter, @johanste approve the PR so I can proceed with merge.
Thank you

dzeliar committed:
bump up version

dzeliar committed:
Merge branch 'main' into dusanzeliar/trunkstatus-2022-07-01

dzeliar committed:
bump up the swagger version

dzeliar committed:
fix errors missing enabled property

dzeliar committed:
revert error code version back to new

mikekistler commented:
/azp run

dzeliar committed:
fix typo and duplicate definition in previous reamdme versions

dzeliar committed:
fix typo

stewartadam commented on specification/communication/data-plane/SipRouting/preview/2023-01-01-preview/examples/GetSipConfigurationWithExpand.json:
```diff
@@ -0,0 +1,68 @@
+{
+  "parameters": {
+    "api-version": "2023-01-01-preview",
+    "endpoint": "https://resourcename.communication.azure.com/"
```
I think expand is missing here?

dzeliar replied:
Yes, thanks you. Added it there

dzeliar committed:
address comments

dzeliar committed:
update ping description

dzeliar commented:
Hello everyone,

on yesterdays board review of SipRouting - Trunk health, https://github.com/Azure/azure-rest-api-specs/issues/19376, we agreed on the following steps:

1.  You should not create new schemas for Expanded SipConfiguration or ExpandedTrunk. Just add "health" as an optional property the the existing Trunk schema. You can mark health as readOnly: true for documentation purposes.
2.  Don’t use the “$” prefix character on $expand
3.  Enum value for expand should be "trunks/health"
4. Health should be a named schema that is $ref'ed from here. And any objects nested within Health should be described by separate named schemas -- not nested inline.

[OpenAPI diff](https://portal.azure-devex-tools.com/tools/diff/eyJzb3VyY2VVcmwiOiJodHRwczovL2dpdGh1Yi5jb20vQXp1cmUvYXp1cmUtcmVzdC1hcGktc3BlY3MvYmxvYi81NzUyMDEzOGNlNTk1YzhhZGI3YTA1OTBkZGU1YzJmNTc1MDJjNzA1L3NwZWNpZmljYXRpb24vY29tbXVuaWNhdGlvbi9kYXRhLXBsYW5lL1NpcFJvdXRpbmcvcHJldmlldy8yMDIyLTEwLTAxLXByZXZpZXcvY29tbXVuaWNhdGlvbnNlcnZpY2Vzc2lwcm91dGluZy5qc29uIiwidGFyZ2V0VXJsIjoiaHR0cHM6Ly9naXRodWIuY29tL0F6dXJlL2F6dXJlLXJlc3QtYXBpLXNwZWNzL2Jsb2IvNTc1MjAxMzhjZTU5NWM4YWRiN2EwNTkwZGRlNWMyZjU3NTAyYzcwNS9zcGVjaWZpY2F0aW9uL2NvbW11bmljYXRpb24vZGF0YS1wbGFuZS9TaXBSb3V0aW5nL3ByZXZpZXcvMjAyMy0wMS0wMS1wcmV2aWV3L2NvbW11bmljYXRpb25zZXJ2aWNlc3NpcHJvdXRpbmcuanNvbiJ9) 

I addressed all of the suggestions above. The changes were pushed, now I need an approval from API Stewardship Board.
Please, @mikekistler , @JeffreyRichter, @DominikMe review and approve the PR so it can be merged.
Thank you

dzeliar committed:
shuffle object definition to help APIview render document properly

dzeliar committed:
Revert "shuffle object definition to help APIview render document properly"

This reverts commit a682d6a26f93866ca3467253168be5c569e958a6.

mikekistler reviewed (approved):
Looks good. 👍 

I flagged a few minor issues.

weidongxu-microsoft reviewed (approved):
LGTM.

Same minor comment about redundant "type: object" on "$ref"

dzeliar committed:
remove redundant type:object on

dzeliar commented on specification/communication/data-plane/SipRouting/preview/2023-01-01-preview/communicationservicessiprouting.json:
```diff
@@ -0,0 +1,573 @@
+        },
+        "overall": {
+          "type": "object",
+          "description": "The overall health status of Trunk.",
```
Updated

dzeliar commented:
Thank you for reviews @mikekistler , @weidongxu-microsoft . As merging the PR is restricted to authorized users only, can any of the code-owners perform merge? cc @DominikMe , @scgbear , @JeffreyRichter

mikekistler commented:
@lirenhe Could you please merge this PR.
